File: jx/im_client/plugin/flutter_yun_ceng_kiwi/ios/Classes/FlutterYunCengKiwiPlugin.m
#import "FlutterYunCengKiwiPlugin.h"
// #include <YunCeng/YunCeng.h>
#import <Kiwi/Kiwi.h>
#import <Flutter/Flutter.h>

const int kiwi_init_default = -999;
int kiwi_init_value = kiwi_init_default;

@implementation FlutterYunCengKiwiPlugin{
    NSString* _token;
    FlutterMethodChannel *_methodChannel;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"flutter_yun_ceng_kiwi"
            binaryMessenger:[registrar messenger]];
    FlutterYunCengKiwiPlugin* instance = [[FlutterYunCengKiwiPlugin alloc] initWithChannel:channel];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithChannel:(FlutterMethodChannel *)channel {
    self = [super init];
    if (self) {
        _methodChannel = channel;
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"getPlatformVersion" isEqualToString:call.method]) {
        result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    }
    else if ([call.method isEqualToString:@"initEx"]) {
        return [self initEx:call result:result];
    }
    else if ([call.method isEqualToString:@"initAsync"]) {
        return [self initAsync:call result:result];
    }
    else if ([call.method isEqualToString:@"isInitDone"]) {
        return [self isInitDone:call result:result];
    }
    // else if ([call.method isEqualToString:@"initExWithCallback"]) {
    //     return [self initExWithCallback:call result:result];
    //  }
    else if ([call.method isEqualToString:@"getProxyTcpByDomain"]) {
        return [self getProxyTcpByDomain:call result:result];
    }
    else if([call.method isEqualToString:@"restartAllServer"]){
        return [self restartAllServer:call result:result];
    }
    else if([call.method isEqualToString:@"onNetworkOn"]){
        return [self onNetworkOn:call result:result];
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

- (void) restartAllServer:(FlutterMethodCall*)call result:(FlutterResult)result
{
    [Kiwi RestartAllServer];
    NSLog(@"游戏盾网络重置成功");
    result(nil);
}

- (void) onNetworkOn:(FlutterMethodCall*)call result:(FlutterResult)result
{
    [Kiwi OnNetworkOn];
    NSLog(@"游戏盾网络恢复成功");
    result([NSNumber numberWithInt:0]);
}

- (void) initEx:(FlutterMethodCall*)call result:(FlutterResult)result
{
    NSDictionary *args = call.arguments;
    _token = args[@"token"];
    NSString* appKey = args[@"appKey"];
    int ret = [Kiwi Init:[appKey UTF8String]];
    kiwi_init_value = ret;
    if (0 != ret) {
        NSLog(@"初始化游戏盾失败:%i, 准备重试!",ret);
    }else{
        NSLog(@"yxd初始化成功%i",ret);
    }
    result([NSNumber numberWithInt:ret]);
}

- (void) initAsync:(FlutterMethodCall*)call result:(FlutterResult)result
{
    NSDictionary *args = call.arguments;
    _token = args[@"token"];
    NSString* appKey = args[@"appKey"];
    kiwi_init_value = kiwi_init_default;
    [Kiwi InitWithListener:[appKey UTF8String] :^(int code){
        kiwi_init_value = code;
    }];
    result([NSNumber numberWithInt:0]);
}

- (void) isInitDone:(FlutterMethodCall*)call result:(FlutterResult)result
{
    if (kiwi_init_value != kiwi_init_default) {
        result([NSNumber numberWithInt:0]);
    } else {
        result([NSNumber numberWithInt:-1]);
    }
}

// - (void) initExWithCallback:(FlutterMethodCall*)call result:(FlutterResult)result
// {
//     NSDictionary *args = call.arguments;
//     _token = args[@"token"];
//     NSString* appKey = args[@"appKey"];
    
//     __weak typeof(self) weakSelf = self;
//     [YunCeng initExWithCallback:[appKey UTF8String] :[_token UTF8String] :^(int ret){
//         __strong typeof(self) strongSelf = weakSelf;
//         if (!strongSelf) {
//             return ;
//         }
        
//         if (0 != ret) {
//             NSLog(@"初始化游戏盾失败:%i, 准备重试!",ret);
//         }else{
//             NSLog(@"yxd初始化成功%i",ret);
//         }
//         [strongSelf->_methodChannel invokeMethod:@"initExWithCallbackResult" arguments:[NSNumber numberWithInt:ret]];
//     }];
    
//     result(nil);
// }

-(void) getProxyTcpByDomain:(FlutterMethodCall*)call result:(FlutterResult)result
{
//    if (!_token) {
//        NSLog(@"游戏盾未初始化!");
//        return;
//    }
    
    NSDictionary *args = call.arguments;
    NSString* token = _token;
    NSString* groupName = args[@"group_name"];
    NSString* ddomain = args[@"ddomain"];
    NSString* dPort = args[@"dport"];
    
    char ip[128]= {0};
    char port[40] = {0};
    int ret = [Kiwi ServerToLocal :[groupName UTF8String] : ip : sizeof(ip) : port:sizeof(port)];
    
    NSMutableDictionary * dic = [[NSMutableDictionary alloc] init];
    [dic setObject: [NSString stringWithFormat:@"%i",ret] forKey:@"code"];
    [dic setObject: [NSString stringWithUTF8String:ip] forKey:@"target_ip"];
    [dic setObject: [NSString stringWithUTF8String:port] forKey:@"target_port"];
    result(dic);
}


-(BOOL)isNull:(NSMutableDictionary *)dict key:(NSString*)key{
    // judge nil
    if(![dict objectForKey:key]){
        return NO;
    }
    id obj = [dict objectForKey:key];// judge NSNull
    
    BOOL isNull = [obj isEqual:[NSNull null]];
    return isNull;
}

@end

File: lucky/im-client/plugin/flutter_yun_ceng_kiwi/macos/Classes/FlutterYunCengKiwiPlugin.swift
import Cocoa
import FlutterMacOS

let kiwi_init_default : Int32 = 999;
var kiwi_init_value : Int32 = kiwi_init_default;

typealias SwiftInitCallback = @convention(c) (Int32) -> Void

func initCallback(result: Int32) {
    print("Kiwi init with value: \(result)")
    kiwi_init_value = result;
}

public class FlutterYunCengKiwiPlugin: NSObject, FlutterPlugin {
    private var _token: String?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_yun_ceng_kiwi", binaryMessenger: registrar.messenger)
        let instance = FlutterYunCengKiwiPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)
        case "initEx":
            self.initExWith(methodCall: call, result: result)
        case "initAsync":
            self.initAsync(methodCall: call, result: result)
        case "isInitDone":
            self.isInitDone(methodCall: call, result: result)
        case "getProxyTcpByDomain":
            self.getProxyTcpByDomain(methodCall: call, result: result)
        case "onNetworkOn":
            self.onNetworkOn(methodCall: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func initExWith(methodCall: FlutterMethodCall, result: FlutterResult){
        guard let arg = methodCall.arguments as? Dictionary<String, Any> else {
            result(FlutterError.init(code: "-5", message: "bad args", details: nil))
            return;
        }

        if let appKey = arg["appKey"] as? String,
        let token = arg["token"] {
            _token = token as? String
            let res = KiwiInit(appKey.UTF8CString);
            kiwi_init_value = res;
            result(res)
            return;
        }
        
        result(FlutterError.init(code: "-2", message: "app key is nul", details: nil))
        return;
    }

    public func initAsync(methodCall: FlutterMethodCall, result: FlutterResult){
        guard let arg = methodCall.arguments as? Dictionary<String, Any> else {
            result(FlutterError.init(code: "-5", message: "bad args", details: nil))
            return;
        }

        kiwi_init_value = kiwi_init_default;
        if let appKey = arg["appKey"] as? String,
        let token = arg["token"] {
            _token = token as? String
            let callback: SwiftInitCallback = initCallback
            let res = KiwiInitWithListner(appKey.UTF8CString, callback);
            kiwi_init_value = res;
            result(0)
            return;
        }
        
        result(FlutterError.init(code: "-2", message: "app key is nul", details: nil))
        return;
    }

    public func isInitDone(methodCall: FlutterMethodCall, result: FlutterResult){
        if (kiwi_init_value != kiwi_init_default) {
            result(0);
        } else {
            result(1);
        }
    }
    
    public func getProxyTcpByDomain(methodCall: FlutterMethodCall, result: FlutterResult) {
        if _token == nil {
            result(FlutterError.init(code: "-1", message: "游戏盾未初始化!", details: nil))
        }
        
        guard let arg = methodCall.arguments as? Dictionary<String, Any> else {
            result(FlutterError.init(code: "-5", message: "bad args", details: nil))
            return;
        }

        if let domain = arg["group_name"] as? String {
            
            let ip: [UInt8] = Array(repeating: 0, count: 128);
            let port: [UInt8] = Array(repeating: 0, count: 40);
            
            
            let ipString = String(decoding: ip, as: UTF8.self)
            let portString =  String(decoding: port, as: UTF8.self)
            
            let ret = KiwiServerToLocal(domain.UTF8CString, ipString.UTF8CString, Int32(ip.count), portString.UTF8CString, Int32(port.count))
            
            var dict: Dictionary = Dictionary<String, Any>()
            if(String(cString: portString.toCString()).isEmpty == false){
                dict["target_port"] = String(cString: portString.toCString())
            }
            
            if(String(cString: ipString.toCString()).isEmpty == false){
                dict["target_ip"] = String(cString: ipString.toCString())
            }
            
            dict["code"] = ret
            result(dict)
            return;
        }
        
        result(FlutterError.init(code: "-1", message: "arg is nil", details: nil))
        return;
    }

    public func onNetworkOn(methodCall: FlutterMethodCall, result: FlutterResult){
        KiwiOnNetworkOn();
        result(0);
    }
}

extension String{
    
    // String 转化为  char *
    func toCString(usig encoding: String.Encoding = .utf8)->UnsafePointer<CChar>{
        let data = self.data(using: encoding)!
        let bytes = UnsafeMutablePointer<UInt8>.allocate(capacity: data.count)
        data.copyBytes(to: bytes, count: data.count)
        return UnsafeRawPointer(bytes).assumingMemoryBound(to: CChar.self)
    }
    
    var UTF8CString: UnsafeMutablePointer<Int8> {
        return UnsafeMutablePointer<Int8>(mutating: (self as NSString).utf8String!)
    }
    
}
